formats/cbor/commonMain/src/kotlinx/serialization/cbor/internal/Encoding.kt

```diff
@@ -233,10 +233,29 @@ internal open class CborReader(private val cbor: Cbor, protected val decoder: Cb
     }
 
     override fun decodeElementIndex(descriptor: SerialDescriptor): Int {
-        if (!finiteMode && decoder.isEnd() || (finiteMode && readProperties >= size)) return CompositeDecoder.DECODE_DONE
-        val elemName = decoder.nextString()
-        readProperties++
-        val index = descriptor.getElementIndexOrThrow(elemName)
+        val index = if (cbor.ignoreUnknownKeys) {
+            val knownIndex: Int
+            while (true) {
+                if (isDone()) return CompositeDecoder.DECODE_DONE
```

Comment:
This check could be lifted to the very first line of the method to simplify the control flow and make it slightly easier to read

Reply:
Unfortunately `decoder.skipElement()` (line 248) will skip the current value element and we'll need to check `isDone()` on the next loop to keep the previous short-circuit behavior (when beginning to read the next element value).

I tried lifting `isDone()` line (just to be sure) and it does in fact cause a number of tests to fail.

Would it be better to remove some of the code that is duplicated in the top-most `if` branches by having it always enter the `while` loop?:

```kotlin
override fun decodeElementIndex(descriptor: SerialDescriptor): Int {
    val index: Int

    while (true) {
        if (isDone()) return CompositeDecoder.DECODE_DONE
        val elemName = decoder.nextString()
        readProperties++

        if (cbor.ignoreUnknownKeys) {
            val elementIndex = descriptor.getElementIndex(elemName)
            if (elementIndex == CompositeDecoder.UNKNOWN_NAME) {
                decoder.skipElement()
            } else {
                index = elementIndex
                break
            }
        } else {
            index = descriptor.getElementIndexOrThrow(elemName)
            break
        }
    }

    decodeByteArrayAsByteString = descriptor.isByteString(index)
    return index
}
```

---

I agree that the control flow is a bit difficult to follow in this function, so any other suggestions/strategies to refactor and simplify it a bit are appreciated.